File: geqiangbao_20201117_CommandPattern_01/geqiangbao_20201117_CommandPattern_01/day1_Command/TetrisMachine/Simple1/DynamicCommand.m
//
//  DynamicCommand.m
//  geqiangbao_20201117_CommandPattern_01
//
//  Created by xiaobao on 2020/11/17.
//

#import "DynamicCommand.h"

@interface DynamicCommand ()
@property (nonatomic,copy) DynamicBlock block;
@property (nonatomic,strong) TetrisMachine* tm;

@end

@implementation DynamicCommand


-(instancetype)init:(DynamicBlock)block tm:(TetrisMachine*)tm{
    self = [super init];
    if (self) {
        self.block = block;
        self.tm = tm;
    }
    return self;
}


-(void)execute{
    self.block(self.tm);
}






@end

File: 20.命令模式/geqiangbao_20201204_commandPattern_swift/geqiangbao_20201204_commandPattern_swift/命令模式/2.动态命令/DynamicCommand.swift
//
//  DynamicCommand.swift
//  geqiangbao_20201204_commandPattern_swift
//
//  Created by xiaobao on 2020/12/4.
//

import UIKit

/// 动态命令
class DynamicCommand: CommandProtocol {
    private var computer:MacComputer
    private var block:(MacComputer) ->()
    
    init(computer:MacComputer,block:@escaping (MacComputer) ->()) {
        self.computer = computer
        self.block = block
    }
    
    func execute() {
        self.block(self.computer)
    }
    
    class func createCommand(computer:MacComputer,block:@escaping (MacComputer) ->()) -> CommandProtocol{
        return DynamicCommand(computer: computer, block: block)
    }
}
